janus_server/src/datastore.rs

```diff
@@ -1018,6 +1014,109 @@ impl Transaction<'_> {
 
         Ok(rows)
     }
+
+    /// Fetch an `aggregate_share_jobs` row from the datastore corresponding to the provided
+    /// [`AggregateShareRequest`], or `None` if no such job exists.
+    #[tracing::instrument(skip(self), err)]
+    pub(crate) async fn get_aggregate_share_job_by_request<P, A, E>(
```

Comment:
Please make this take a single generic parameter, which is a `vdaf::Aggregator` (or maybe `vdaf::Vdaf`, if you don't need the additional functionality in the `Aggregator` trait), and state the constraints on the existing parameters in terms of the associated types of the `vdaf::Vdaf`.

`AggregateShareJob`'s generic parameters would be changed similarly, as would `put_aggregate_share_job`'s.

The intended meaning of these generic parameters is unnecessarily unclear, and I am not sure there is an upside to making this so flexible -- the intended use is always in the context of a specific `vdaf::Aggregator`. And we've already gone to the trouble of defining the associated types in libprio, let's make use of them. :-)

Reply:
(I realized this was a preexisting issue, see https://github.com/divviup/janus/pull/130 for an example fix. I think the trickiest bit is that the auto-derived `PartialEq`/`Eq` implementations don't seem to play nicely. Maybe it's a GAT thing?)

Reply:
I think the `#[derive(PartialEq)]` issue is referred to as "perfect derive" in [this roadmap blogpost](https://blog.rust-lang.org/inside-rust/2022/04/04/lang-roadmap-2024.html#looking-forward) FWIW, looks like it's orthogonal to GAT.